### ISSUE TITLE
"Using PyTorch C++ Frontend" TorchModule generator architecture is erroneous

### ISSUE THREAD

**Issue by BynaryCobweb** [pytorch/pytorch#25644]
## 📚 Documentation

I followed the DCGAN tutorial on the page ["Using the PyTorch C++ Frontend"](https://pytorch.org/tutorials/advanced/cpp_frontend.html).

Instead of using the sequential version of the generator I used the TorchModule one, and I realized that the two do not match. Additionally, the TorchModule version of the generator does not work with the discriminator.

Here is the sequential version:
```cpp
nn::Sequential generator(
    // Layer 1
    nn::Conv2d(nn::Conv2dOptions(kNoiseSize, 256, 4)
                   .with_bias(false)
                   .transposed(true)),
    nn::BatchNorm(256),
    nn::Functional(torch::relu),
    // Layer 2
    nn::Conv2d(nn::Conv2dOptions(256, 128, 3)
                   .stride(2)
                   .padding(1)
                   .with_bias(false)
                   .transposed(true)),
    nn::BatchNorm(128),
    nn::Functional(torch::relu),
    // Layer 3
    nn::Conv2d(nn::Conv2dOptions(128, 64, 4)
                   .stride(2)
                   .padding(1)
                   .with_bias(false)
                   .transposed(true)),
    nn::BatchNorm(64),
    nn::Functional(torch::relu),
    // Layer 4
    nn::Conv2d(nn::Conv2dOptions(64, 1, 4)
                   .stride(2)
                   .padding(1)
                   .with_bias(false)
                   .transposed(true)),
    nn::Functional(torch::tanh));
```

The erroneous TorchModule version:
```cpp
struct GeneratorImpl : nn::Module {
  GeneratorImpl()
      : conv1(nn::Conv2dOptions(kNoiseSize, 512, 4)
                  .with_bias(false)
                  .transposed(true)),
        batch_norm1(512),
        conv2(nn::Conv2dOptions(512, 256, 4)
                  .stride(2)
                  .padding(1)
                  .with_bias(false)
                  .transposed(true)),
        batch_norm2(256),
        conv3(nn::Conv2dOptions(256, 128, 4)
                  .stride(2)
                  .padding(1)
                  .with_bias(false)
                  .transposed(true)),
        batch_norm3(128),
        conv4(nn::Conv2dOptions(128, 64, 4)
                  .stride(2)
                  .padding(1)
                  .with_bias(false)
                  .transposed(true)),
        batch_norm4(64),
        conv5(nn::Conv2dOptions(64, 1, 4)
                  .stride(2)
                  .padding(1)
                  .with_bias(false)
                  .transposed(true)) {}

  torch::Tensor forward(torch::Tensor x) {
    x = torch::relu(batch_norm1(conv1(x)));
    x = torch::relu(batch_norm2(conv2(x)));
    x = torch::relu(batch_norm3(conv3(x)));
    x = torch::relu(batch_norm4(conv4(x)));
    x = torch::tanh(conv5(x));
    return x;
  }

  nn::Conv2d conv1, conv2, conv3, conv4, conv5;
  nn::BatchNorm batch_norm1, batch_norm2, batch_norm3, batch_norm4;
};
```

And the fixed version I used:
```cpp
struct GeneratorImpl : nn::Module {
  GeneratorImpl(int kNoiseSize)
      : conv1(nn::Conv2dOptions(kNoiseSize, 256, 4)
                  .with_bias(false)
                  .transposed(true)),
        batch_norm1(256),
        conv2(nn::Conv2dOptions(256, 128, 3)
                  .stride(2)
                  .padding(1)
                  .with_bias(false)
                  .transposed(true)),
        batch_norm2(128),
        conv3(nn::Conv2dOptions(128, 64, 4)
                  .stride(2)
                  .padding(1)
                  .with_bias(false)
                  .transposed(true)),
        batch_norm3(64),
        conv4(nn::Conv2dOptions(64, 1, 4)
                  .stride(2)
                  .padding(1)
                  .with_bias(false)
                  .transposed(true)) 
  {
    // register_module() is needed if we want to use the parameters() method later on
    register_module("conv1", conv1);
    register_module("conv2", conv2);
    register_module("conv3", conv3);
    register_module("conv4", conv4);
    register_module("batch_norm1", batch_norm1);
    register_module("batch_norm2", batch_norm1);
    register_module("batch_norm3", batch_norm1);
  }

  torch::Tensor forward(torch::Tensor x) {
    x = torch::relu(batch_norm1(conv1(x)));
    x = torch::relu(batch_norm2(conv2(x)));
    x = torch::relu(batch_norm3(conv3(x)));
    x = torch::tanh(conv4(x));
    return x;
  }

  nn::Conv2d conv1, conv2, conv3, conv4;
  nn::BatchNorm batch_norm1, batch_norm2, batch_norm3;
};



cc @ezyang @gchanan @zou3519 @yf225

**Comment by soumith** [comment]
cc: @jlin27 @ShahriarSS @yf225 for fixing this

**Comment by ailzhang** [comment]
I removed high pri label and added this to 1.3 milestone instead.

**Comment by jlin27** [comment]
Corrected errors in TorchModule version: https://github.com/pytorch/tutorials/pull/645

**Comment by BynaryCobweb** [comment]
Almost right... but there is no 5th convolution layer:
```cpp
// ...
        batch_norm3(64),
        conv4(nn::Conv2dOptions(64, 1, 4)
                  .stride(2)
                  .padding(1)
                  .with_bias(false)
                  .transposed(true))
// ----------------------------
        batch_norm4(64),
        conv5(nn::Conv2dOptions(64, 1, 4)
                  .stride(2)
                  .padding(1)
                  .with_bias(false)
                  .transposed(true))
// -------- this should be removed ^^^^^
{
```

**Comment by jlin27** [comment]
@BynaryCobweb - Good catch. Removed the erroneous remaining conv5: https://github.com/pytorch/tutorials/pull/664

**Comment by gchanan** [comment]
@yf225 can you make sure this gets pushed through for the release?

**Comment by yf225** [comment]
@gchanan : Based on conversation with @jlin27, this is pushed through for the v1.3.0 release as well.
